changefeedccl: add support for IAM role-based authentication to Kafka sink

cty123 opened the pull request:
This patch adds support to the go Kafka client(Sarama) to authenticate
with AWS IAM roles. AWS IAM role authentication is the recommended
way to connect with AWS MSK cluster as it's more secure than the
SASL username + password method.

AWS now provides a library in a repository under their official account
https://github.com/aws/aws-msk-iam-sasl-signer-go that is able to
connect to AWS MSK with IAM roles. This PR introduces the new
dependency on the library and provides option in Kafka DialConfig
to enable the IAM role authentcaition.

Resolves: #124431

Release note (enterprise change): Add support to the Kafka sink
to authenticate with MSK via AWS IAM roles

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/125745)
<!-- Reviewable:end -->

rharding6373 reviewed (commented):
Thanks for the contribution @cty123! This is very helpful. I did a first pass and only have a minor nit. @andyyang890 is going to mirror the libraries and test the feature.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125745#-:-O-pyjTXCGk_7E-_preI:begx7q5)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890 and @cty123)

___
*[`pkg/ccl/changefeedccl/sink_kafka.go` line 789 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125745#-O-b0ZLZBE-q3JB8jOzw:-O-b0ZLZBE-q3JB8jOzx:b9kwnnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc5fca5fac71cd5384cccf396e57a36c0d9628a/pkg/ccl/changefeedccl/sink_kafka.go#L789)):*
> ```Go
> }
> 
> type AwsIamRoleSaslTokenProvider struct {
> ```

nit: replace `Iam`with all-caps `IAM` throughout (keeping `iam` when at the beginning is ok).
___
*[`pkg/ccl/changefeedccl/sink_kafka_connection_test.go` line 220 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125745#-O-b5EYt6C2oeex_ZGPc:-O-b5EYt6C2oeex_ZGPd:b-d922oe) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dc5fca5fac71cd5384cccf396e57a36c0d9628a/pkg/ccl/changefeedccl/sink_kafka_connection_test.go#L220)):*
> ```Go
> 		},
> 		{
> 			name:          "sasl_aws_iam_session_name must be provided when AWS IAM authentication is enabled",
> ```

Could you please add another test that omits `sasl_mechanism`?


<!-- Sent from Reviewable.io -->

asg0451 reviewed (changes requested):
Thanks for the contribution @cty123! Overall this looks good and I can confirm that it works on my test MSK clusters. I have one nitpick about the params you added, and pls remove the `c-deps` file that I assume got accidentally included. Then we'll be good to go!

Please feel free to @ me directly if you need any help.

cty123 commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -219,6 +219,10 @@ const (
 	SinkParamSASLTokenURL           = `sasl_token_url`
 	SinkParamSASLScopes             = `sasl_scopes`
 	SinkParamSASLGrantType          = `sasl_grant_type`
+	SinkParamSASLAwsIAMRoleEnabled  = `sasl_aws_iam_role_enabled`
```
I am definitly for it, the only concern that prevented from doing that was the fact this field is currently using a set of constants provided from sarama library: 

```go
const (
	// SASLTypeOAuth represents the SASL/OAUTHBEARER mechanism (Kafka 2.0.0+)
	SASLTypeOAuth = "OAUTHBEARER"
	// SASLTypePlaintext represents the SASL/PLAIN mechanism
	SASLTypePlaintext = "PLAIN"
	// SASLTypeSCRAMSHA256 represents the SCRAM-SHA-256 mechanism.
	SASLTypeSCRAMSHA256 = "SCRAM-SHA-256"
	// SASLTypeSCRAMSHA512 represents the SCRAM-SHA-512 mechanism.
	SASLTypeSCRAMSHA512 = "SCRAM-SHA-512"
	SASLTypeGSSAPI      = "GSSAPI"
	// SASLHandshakeV0 is v0 of the Kafka SASL handshake protocol. Client and
	// server negotiate SASL auth using opaque packets.
	SASLHandshakeV0 = int16(0)
	// SASLHandshakeV1 is v1 of the Kafka SASL handshake protocol. Client and
	// server negotiate SASL by wrapping tokens with Kafka protocol headers.
	SASLHandshakeV1 = int16(1)
	// SASLExtKeyAuth is the reserved extension key name sent as part of the
	// SASL/OAUTHBEARER initial client response
	SASLExtKeyAuth = "auth"
)
```

and in Sarama client, we still need to map `sasl_mechanism=AWS_MSK_IAM` back to `sarama.SASLTypeOAuth`, given the usage example provided by the AWS library, https://github.com/aws/aws-msk-iam-sasl-signer-go

```go
// Set the SASL/OAUTHBEARER configuration
config := sarama.NewConfig()
config.Net.SASL.Enable = true
config.Net.SASL.Mechanism = sarama.SASLTypeOAuth
config.Net.SASL.TokenProvider = &MSKAccessTokenProvider{}
...
```

I was a bit afraid that this could make the logic more complicated

cty123 replied:
I have added an additional option for the SASL mechanism called `sasl_mechanism=AWS_IAM_ROLE` to be more generic, since the server doesn't always have to be MSK. Used the name `IAM_ROLE`, in case there is support for `IAM USER` in the future, even though using IAM user is a bad idea for most use cases.

asg0451 replied:
@cty123 you're right about having to do this behind sarama's back a bit, but i still think it would be worth it. Would you mind using `AWS_MSK_IAM`? my reasoning is as follows:

- iam auth is indeed specific to msk
- it's what amazon's sasl plugin uses (see step 8 of https://docs.aws.amazon.com/msk/latest/developerguide/create-topic.html) - 
![image](https://github.com/cockroachdb/cockroach/assets/4358545/be639226-f825-4d70-968d-5741169ca38e)
- we will want to support non-role auth in the future, and the way we'll do that is by using the same mechanism name and omitting the `sasl_aws_iam_role_arn` parameter

can you update this pr to reflect the above pls?

cty123 replied:
I guess that makes sense, I wasn't aware of this AWS page.

cty123 commented on c-deps/libedit:
Probably introduced from git rebase, will clean this up

cty123 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -785,6 +786,38 @@ func (t *tokenProvider) Token() (*sarama.AccessToken, error) {
+
+	if len(dialConfig.saslAwsIAMRoleArn) == 0 {
+		return nil, errors.New("saslAwsIamRole can not be empty")
+	}
```
Removed the validation logic. I think you are right, when the parameter is consumed, this empty check should have already happened

cty123 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -874,21 +907,25 @@ func getSaramaConfig(
+	saslClientSecret      string
+	saslScopes            []string
+	saslGrantType         string
+	saslAwsIAMRoleEnabled bool
```
Removed

asg0451 commented on pkg/ccl/changefeedccl/sink_kafka.go:
```diff
@@ -785,6 +786,30 @@ func (t *tokenProvider) Token() (*sarama.AccessToken, error) {
+func (p *AwsIAMRoleSaslTokenProvider) Token() (*sarama.AccessToken, error) {
+	token, _, err := signer.GenerateAuthTokenFromRole(
+		p.ctx, p.awsRegion, p.iamRoleArn, p.iamSessionName)
+	return &sarama.AccessToken{Token: token}, err
```
nit: return `nil, err` or `&sarama.AccessToken{..}, nil`

asg0451 commented on pkg/ccl/changefeedccl/changefeedbase/options.go:
```diff
@@ -239,6 +242,9 @@ const (
 )
 
+// Support additional mechanism on top of the default SASL mechanism.
+const SASLTypeAwsIAMRole = "AWS_MSK_IAM"
```
nit: change the name of the constant too pls

cty123 replied:
Nice catch, I forgot that, btw I don't fully understand the casing convention in the names, because Aws is written with capitalized A, and IAM, SASL(and MSK I guess) are written as all caps. Meanwhile, GCP is in all caps, not sure if `SASLTypeAwsMSKIAM` is complaint to the CRL naming convention.

asg0451 replied:
@cty123 I think the most correct name would be `SASLTypeAWSMSKIAM` but that's kind of hard to read. but now that you mention it, why dont you go ahead and change it to that ;)

cty123 replied:
Are you sure about that, because I see in other place AWS is always written as `Aws` like here: https://github.com/cockroachdb/cockroach/blob/46a5cb142d7f9c19c190d63dc7c8f91195e4ef3c/pkg/roachprod/vm/aws/aws.go#L59

cty123 replied:
I changed to `SASLTypeAWSMSKIAM` anyways, I see both casing being used in code

asg0451 reviewed (approved):
@cty123 looks good! just a few nits in the code, and can you change the commit msg as well? thanks for following the guidelines in https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73072807/Git+Commit+Messages - i just have a few requests:
- can you add `Resolves: #124431` to your message
- for the release note, can you say `Release note (enterprise change): Add support to the Kafka sink to authenticate with MSK via AWS IAM roles`

thanks!

cty123 commented:
For some reason previously the library `github.com/aws/aws-sdk-go-v2/service/sts v1.19.3` was commented as indirect that caused gitlab workflow failure, fixed that in the latest push. https://github.com/cockroachdb/cockroach/actions/runs/9860634735/job/27229521713

asg0451 commented on c-deps/libedit:
@cty123 seems like this found its way back in, which is why CI is failing

cty123 replied:
Well, at least I found why I got that in, my vscode by default includes all file changes. libedit used to be in the c-deps repo, and then it was dropped(which was pretty nice improvement), and I didn't delete it on my local workspace.

asg0451 replied:
thanks @cty123 . I asked @rharding6373 to have one final look and then when CI passes i'll merge this for you. thanks again for the contribution!

cty123 committed:
changefeedccl: add support for IAM role-based authentication to Kafka sink

This patch adds support to the go Kafka client(Sarama) to authenticate
with AWS IAM roles. AWS IAM role authentication is the recommended
way to connect with AWS MSK cluster as it's more secure than the
SASL username + password method.

AWS now provides a library in a repository under their official account
https://github.com/aws/aws-msk-iam-sasl-signer-go that is able to
connect to AWS MSK with IAM roles. This PR introduces the new
dependency on the library and provides option in Kafka DialConfig
to enable the IAM role authentcaition.

Resolves: #124431
Release note (enterprise change): Add support to the Kafka sink
to authenticate with MSK via AWS IAM roles

rharding6373 reviewed (approved):
LGTM. Thank you for your contribution!

rharding6373 commented:
bors r=asg0451,rharding6373
